### PR TITLE
GroupMembershipResolver: Fix issues with apply and static assignments

### LINKED ISSUE #1574: Integrity constraint violation: icinga_hostgroup_host_resolved
After update from r2.8.4-1 to 2.9, all previous defined assign filters on HostGroups are no longer visible (but active) and an error message appears on saving assign filters on new and existing HostGroups.

## Expected Behavior
No error message should appear and all previous defined assign filters should be there.

## Current Behavior
Error message appears on save:
`SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '15-3' for key 'PRIMARY', query was: INSERT INTO icinga_hostgroup_host_resolved (hostgroup_id, host_id) VALUES (?, ?) (Pdo.php:225)`

## Possible Solution
NA

## Steps to Reproduce (for bugs)
Create/edit a HostGroup and set/edit a assign filter

## Your Environment
* Director version (System - About): master
* Icinga Web 2 version and modules (System - About): 2.5.3
* Icinga 2 version (`icinga2 --version`): r2.9.0-1
* Operating System and version: Ubuntu 18.04 LTS
* Webserver, PHP versions: nginx/1.14.0, php7.2.7-1

### PR DESCRIPTION
**Pull request by lazyfrosch**
* Fixing static resolving in general
* Avoiding a problem where apply were matched when assign_filter is an
  empty string (empty FilterAnd)
* Adapt to ResolverTest

fixes #1574

### CODE REVIEW

**Comment by lazyfrosch** [comment]
This addresses an issue where updates are mixed over the intended range of objects and groups. Mapping are updated where the respective groups are not in scope, and therefor can not be checked for "existing".

GroupMembershipResolver has 3 modes we use:
1. All groups for one object (changing a host)
2. All objects for one group (changing a host group)
3. All for all (e.g. after sync)

We should also add unit tests probably...

**Comment by lazyfrosch** [comment]
Following up on this, my change here just excludes static hostgroups from the resolver.

Working on a better solution, including tests

**Comment by lazyfrosch** [comment]
This has become a quite large diff, but still only bugfixing issues with the resolver.

The Test included should test the basic behavior when changing objects.

@Thomas-Gelf please review!

**Comment by lazyfrosch** [comment]
There seems to be a problem with PHP < 5.6 - need to verify this

**Comment by lazyfrosch** [comment]
Included housekeeping and syncmode from #1490

**Comment by lazyfrosch** [comment]
Removed a mis-commited file from testing

**Comment by lazyfrosch** [comment]
Current testing this in a larger environment.

Still having some problems with group assign -> host when going by vars, and the var is inherited from a template.

Hostgroup has been updated by sync, a housekeeping call didn't fix it, but manual change of the group

**Comment by lazyfrosch** [comment]
Tracked this down to a rogue custom var, mapping seem better now, but still testing.

**Comment by Thomas-Gelf** [comment]
@lazyfrosch: I tried a simplistic approach for just a single use-case, #1574 - could you please have a quick look at 131a921 (temp. branch: fix/hostgroup-membership-resolver)?

**Comment by Thomas-Gelf** [comment]
@lazyfrosch: would splitting this into "single group updates like in #1574", "massive changes in Sync" and "housekeeping utility for group membership cache" cause a problem for your working environment?

**Comment by lazyfrosch** [comment]
Removed all commits that are for testing and housekeeping.

We will add this to 1.6 afterwards

File: library/Director/Objects/GroupMembershipResolver.php [hunk 1]
```diff
@@ -392,8 +431,14 @@ protected function addMembershipWhere(ZfSelect $query, $column, & $objects)
         }
 
         $ids = array();
-        foreach ($objects as $object) {
-            $ids[] = (int) $object->get('id');
+        foreach ($objects as $k => $object) {
+            if (is_string($object)) {
+                $ids[] = (int) $k;
+            } elseif (is_numeric($object)) {
```
Comment (Thomas-Gelf) [inline library/Director/Objects/GroupMembershipResolver.php]: Please change  this to is_int, remove (int) in the line below - then merge to master

Comment (lazyfrosch) [reply library/Director/Objects/GroupMembershipResolver.php]: done

**Comment by lazyfrosch** [comment]
Cherry-Picked to `support/1.5`

**Comment by Thomas-Gelf** [comment]
@lazyfrosch: sh**... discovered a problem: this now writes also matching templates to the DB... was that intentional?

**Pull request merged**